File: engine/src/main/kotlin/app/ledger/engine/Split.kt
package app.ledger.engine

/** Identifies one person within a trip. */
@JvmInline
value class MemberId(val value: String)

/**
 * Divides [totalMinor] equally between [members], in minor currency units.
 *
 * The parts always sum to exactly [totalMinor] — never a cent more or less.
 */
fun splitEqually(totalMinor: Long, members: List<MemberId>, salt: Long = 0): Map<MemberId, Long> =
    shares(totalMinor, members, SplitRule.Equal, salt)

/** How an item's total is divided between the people on it. */
sealed interface SplitRule {
    /** Everyone on the item pays the same. */
    data object Equal : SplitRule

    /** Proportional shares — what the demo's draggable SplitBar produces. */
    data class Weighted(val weights: Map<MemberId, Int>) : SplitRule

    /** Amounts typed against each name; they must add up to the item total. */
    data class Exact(val amounts: Map<MemberId, Long>) : SplitRule
}

/**
 * Divides [totalMinor] between [members] according to [rule].
 *
 * Whichever rule is used, the parts sum to exactly [totalMinor].
 */
fun shares(
    totalMinor: Long,
    members: List<MemberId>,
    rule: SplitRule,
    salt: Long = 0,
): Map<MemberId, Long> {
    require(members.isNotEmpty()) { "cannot split $totalMinor between no one" }
    // Refunds (§9) are a decision not yet taken; until they are, a negative total must be refused
    // here rather than quietly mis-split — Kotlin division truncates toward zero, so the largest
    // remainder arithmetic below is only correct for totals that are not negative.
    require(totalMinor >= 0) { "cannot split a negative amount, got $totalMinor" }
    require(members.size == members.toSet().size) { "the same person cannot be on one item twice" }

    return when (rule) {
        SplitRule.Equal -> {
            splitByWeight(totalMinor, members, members.associateWith { 1 }, salt)
        }

        is SplitRule.Weighted -> {
            val weights = members.associateWith { member ->
                val weight = rule.weights[member] ?: 0
                require(weight >= 0) { "${member.value} has a negative weight of $weight" }
                weight
            }
            require(weights.values.any { it > 0 }) { "at least one person must carry some weight" }
            splitByWeight(totalMinor, members, weights, salt)
        }

        is SplitRule.Exact -> {
            require(rule.amounts.keys == members.toSet()) {
                "exact amounts must name exactly the people on the item"
            }
            // Checked per person, not only in sum: a negative here and padding there cancel out.
            rule.amounts.forEach { (member, amount) ->
                require(amount >= 0) { "${member.value} has a negative amount of $amount" }
            }
            val given = rule.amounts.values.sum()
            require(given == totalMinor) {
                "exact amounts add up to $given but the item total is $totalMinor"
            }
            members.associateWith { rule.amounts.getValue(it) }
        }
    }
}

private class Part(val member: MemberId, val base: Long, val fraction: Long, val rotated: Int)

/**
 * Largest remainder: floor every share, then hand the leftover cents to the largest
 * fractional parts.
 *
 * Ties break on position rotated by [salt], so across a trip's items the spare cent lands
 * on different people rather than always the first name on the list. With every weight at
 * 1 every fraction ties, which makes this reduce exactly to [splitEqually].
 */
private fun splitByWeight(
    totalMinor: Long,
    members: List<MemberId>,
    weights: Map<MemberId, Int>,
    salt: Long,
): Map<MemberId, Long> {
    val totalWeight = weights.values.sumOf { it.toLong() }
    val count = members.size
    val offset = (salt % count).toInt()

    // Refusing beats silently wrapping: past this, totalMinor * weight no longer fits in a Long
    // and every number that falls out of the wreckage still looks plausible. The web port draws
    // the same line at its own limit (Number.MAX_SAFE_INTEGER).
    val heaviest = weights.values.max()
    require(heaviest == 0 || totalMinor <= Long.MAX_VALUE / heaviest) {
        "the amount and weights are too large to split exactly"
    }

    val parts = members.mapIndexed { index, member ->
        val numerator = totalMinor * weights.getValue(member)
        Part(
            member = member,
            base = numerator / totalWeight,
            fraction = numerator % totalWeight,
            rotated = ((index - offset) % count + count) % count,
        )
    }

    // Leftover is always smaller than the number of people carrying a non-zero fraction,
    // so someone on a zero weight can never be handed a stray cent.
    val leftover = (totalMinor - parts.sumOf { it.base }).toInt()
    val getsSpareCent = parts
        .sortedWith(compareByDescending<Part> { it.fraction }.thenBy { it.rotated })
        .take(leftover)
        .mapTo(mutableSetOf()) { it.member }

    return parts.associate { it.member to it.base + if (it.member in getsSpareCent) 1L else 0L }
}

File: web/src/lib/split.ts
/**
 * Largest remainder, in the browser — a second implementation of the one rule this application
 * cannot afford to get wrong.
 *
 * It exists because the SplitBar shows each person's amount live as it is dragged, before anything
 * has been saved. Predicting the server's answer means predicting which member absorbs each spare
 * cent, which depends on the item's id; that is why the client mints the id up front.
 *
 * This is a deliberate duplication of `engine/src/main/kotlin/app/ledger/engine/Split.kt`, and the
 * two are pinned to each other by `split-vectors.json`: the engine regenerates and checks it, and
 * `tests/split.spec.ts` checks this port against the same file. Neither can move without a red
 * test. **If you change anything here, change Split.kt too, or the number somebody sees while
 * dragging will not be the number they are charged.**
 */

/** Beyond this, `total * weight` stops being exact in a double and the arithmetic silently rots. */
const MAX_EXACT = Number.MAX_SAFE_INTEGER

export interface SplitInput {
  totalMinor: number
  /** One weight per person, in the order they appear on the item. Equal split is every weight 1. */
  weights: number[]
  /**
   * The item's id as a 64-bit value. A bigint because that is what it is on the server — reducing
   * it to a double first would change `salt % count` for some member counts and not others, which
   * is the worst kind of wrong.
   */
  salt: bigint
}

export function splitShares({ totalMinor, weights, salt }: SplitInput): number[] {
  const count = weights.length
  if (count === 0) throw new Error('cannot split between no one')
  if (!Number.isInteger(totalMinor)) throw new Error(`total must be whole minor units, got ${totalMinor}`)
  if (weights.some((w) => !Number.isInteger(w) || w < 0)) {
    throw new Error('every weight must be a whole number that is not negative')
  }

  const totalWeight = weights.reduce((a, b) => a + b, 0)
  if (totalWeight === 0) throw new Error('at least one person must carry some weight')
  if (Math.abs(totalMinor) * Math.max(...weights) > MAX_EXACT) {
    // Refusing beats quietly returning a number that is a cent or two out.
    throw new Error('amount and weights are too large to split exactly')
  }

  // Done in bigint and only then narrowed: `count` is small, so the result always fits, but the
  // salt itself does not. Kotlin's `%` keeps the sign of the salt and so does bigint's, which is
  // what makes the rotation below agree with the engine for negative salts too.
  const offset = Number(salt % BigInt(count))

  const parts = weights.map((weight, index) => {
    const numerator = totalMinor * weight
    return {
      index,
      base: Math.floor(numerator / totalWeight),
      fraction: numerator % totalWeight,
      rotated: (((index - offset) % count) + count) % count,
    }
  })

  const distributed = parts.reduce((sum, part) => sum + part.base, 0)
  const leftover = totalMinor - distributed

  // Biggest fractional part first; ties broken by position rotated by the salt, so the spare cent
  // lands on different people across a trip rather than always the first name on the list.
  const winners = new Set(
    [...parts]
      .sort((a, b) => b.fraction - a.fraction || a.rotated - b.rotated)
      .slice(0, leftover)
      .map((part) => part.index),
  )

  return parts.map((part) => part.base + (winners.has(part.index) ? 1 : 0))
}

/**
 * The salt for an item, from its UUID — the browser's copy of `engineItemId` on the server.
 *
 * Both halves are folded together with XOR. Taking only the high 64 bits, as this once did, means
 * two ids differing solely in their low bits share a salt — and on the server, an identity. That
 * is exactly what UUIDv7 produces for two items created in the same millisecond.
 */
export function saltFor(itemId: string): bigint {
  const hex = itemId.replace(/-/g, '')
  if (!/^[0-9a-fA-F]{32}$/.test(hex)) throw new Error(`not a uuid: ${itemId}`)

  const unsigned = BigInt(`0x${hex.slice(0, 16)}`) ^ BigInt(`0x${hex.slice(16)}`)
  // Two's complement, exactly as Java reads a Long.
  return unsigned >= 1n << 63n ? unsigned - (1n << 64n) : unsigned
}

/** A v4 UUID for a new item, so its salt — and therefore its rounding — is known before saving. */
export function newItemId(): string {
  return crypto.randomUUID()
}
